OPSEXP-1690: Document pipenv usage for developers (#411)
Alfresco/alfresco-ansible-deployment: docs/developer-guide.md

## Basic pipenv knowledge

The general purpose of pipenv is similar to that of Package installer for Python (PIP) and built-in venv.
This is external python library, which handles package installing from either commandline (with special pipenv prefix in shell, similar to pip install), requirements.txt or Pipfile.
Whether we install one package or dependencies from requirements.txt, pipenv creates Pipfile, a special file for itself,
and then installs packages we specified inside virtual environment, NOT globally.
While specifyng new package, pipenv adds that package name to Pipfile, installs it (while also generating hashcodes for .lock file)
and installs it inside virtual environment. Then we can also use this package we have installed.
The default virtual environment is created in our working directory. You may want to visit: <https://pipenv.pypa.io/en/latest/>

> NOTE: Pipenv does not install packages globally but to virtual environment

## Basic pipenv commands and usage for developers

Pipenv comes with bunch of commands, but the most important are highlighted below:

> The command with flag --dev installs packages from Pipfile needed for developing purposes.

Reviewer: ```suggestion
The following command will install packages from Pipfile needed for developing purposes:
```

I think it's already pretty self-explanatory (there only one option in the command)

```bash
pipenv install --dev
```

> This command below is opening the virtual environment, that pipenv created while installing all packages.

Reviewer: ```suggestion
The below command is for opening a shell inside the virtual environment that pipenv created previously.
```

> Once we are inside this environment, we can use all the packages which we specified in install command.

Reviewer: ```suggestion
Once we are inside this shell, we can use all the packages that have been installed in the virtual environment.
```

> This is the environment where it is highly recommended to develop python.

Reviewer: ```suggestion
This is the recommended way while actively developing with this repository.
```

```bash
pipenv shell
```

> NOTE: We suggest for developers to run command pipenv shell and then use commands normally, within virtual environment

> > for example: molecule test

Reviewer: this is not something we already stated in the previous paragraph?

> Otherwise, if you are not planning to use virtual environment and need to just simply use molecule (or any other package),

Reviewer: ```suggestion
Otherwise, if you need to just run a few commands inside the virtual environment (e.g. run a molecule test),
```

> you can use these commands which will execute them inside pipenv's virtual environment:

Reviewer: ```suggestion
you can prefix your commands as in the below example which will execute command inside the virtual environment:
```

# Developers' guide

> This page describes how to use pipenv library and pipenv commands when you are developing on this repository.

Reviewer 2: maybe edit this overview to reflect that this page is not only about pipenv but also general development including roles and integration tests

Author: Yes, you are right, that was a leftover since it was created as developer's guide to pipenv

## Basic pipenv knowledge

> The general purpose of pipenv is similar to that of Package installer for Python (PIP) and built-in venv.

Reviewer: link the official docs https://pipenv.pypa.io/en/latest/

Author: Link at the end of the paragraph

This is external python library, which handles package installing from either commandline (with special pipenv prefix in shell, similar to pip install), requirements.txt or Pipfile.
Whether we install one package or dependencies from requirements.txt, pipenv creates Pipfile, a special file for itself,
and then installs packages we specified inside virtual environment, NOT globally.
While specifyng new package, pipenv adds that package name to Pipfile, installs it (while also generating hashcodes for .lock file)
and installs it inside virtual environment. Then we can also use this package we have installed.
The default virtual environment is created in our working directory.

> NOTE: Pipenv does not install packages globally but to virtual environment

## Basic pipenv commands and usage for developers

Pipenv comes with bunch of commands, but the most important are highlighted below:

```bash
pipenv install --dev
```

The command with flag --dev installs packages from Pipfile needed for developing purposes.
If you are not inside the directory where the Pipfile exists, it will create blank Pipfile and not install any package.

> If you plan to work in another directory with the provided packages in Pipfile, you will have to copy that Pipfile to target directory.

Reviewer: I prefer to structure docs containing commands with the explanation before the command.

Otherwise you risk that the user execute the command before reading and understanding what the command do and discover later what was the reason of doing that.

> If you are not inside the directory where the Pipfile exists, it will create blank Pipfile and not install any package.
> If you plan to work in another directory with the provided packages in Pipfile, you will have to copy that Pipfile to target directory.

Reviewer: ```suggestion
```

it works if you are in a subdirectory with an upper Pipfile, I think we can omit these two lines

```bash
pipenv shell
```

This command is opening the virtual environment, that pipenv created while installing all packages.
Once we are inside this environment, we can use all the packages which we specified in install command.
This is the environment where it is highly recommended to develop python.

> NOTE: We suggest for developers to run command pipenv shell and then use commands normally, within virtual environment
> for example: molecule test

Otherwise, if you are not planning to use virtual environment and need to just simply use molecule (or any other package),
you can use these commands which will execute them inside pipenv's virtual environment:

```bash

> pipenv run package_name

Reviewer: ```suggestion
pipenv run command
```

it's really a wrapper for commands available in packages, that the molecule package provide a molecule command is just a molecule characteristic
